### PR TITLE
 A variety of displayio improvements

### LINKED ISSUE #1172: Add text support
Support text rendering from a font builtin.

### LINKED ISSUE #1191: Add subscr to Bitmap
So individual pixels can be read and written.

```python
bitmap[0,0] = 1
```

### LINKED ISSUE #1518: insert/delete into displayio Group
more listy like Groups :)

### PR DESCRIPTION
**Pull request by tannewt**
This changes a number of things in displayio:
* Introduces BuiltinFont and Glyph so the built in font can be used by libraries. For boards with
  a font it is available as board.TERMINAL_FONT. Fixes #1172
* Remove _load_row from Bitmap in favor of bitmap[] access. Index can be x/y tuple or overall index. Fixes #1191
* Add width and height properties to Bitmap.
* Add insert and [] access to Group. Fixes #1518
* Add index param to pop on Group.
* Terminal no longer takes unicode character info. It takes a BuiltinFont instead.
* Fix Terminal's handling of [###D vt100 commands used when up arrowing into repl history.
* Add x and y positions to Group plus scale as well.
* Add bitmap accessor for BuiltinFont

### CODE REVIEW

File: shared-bindings/displayio/Group.c [hunk 1]
```diff
@@ -41,57 +41,236 @@
...
 //|
-//| .. warning:: This will be changed before 4.0.0. Consider it very experimental.
-//|
-//| .. class:: Group(*, max_size=4)
+//| .. class:: Group(*, max_size=4, scale=1)
 //|
 //|   Create a Group of a given size.
 //|
 //|   :param int max_size: The maximum group size.
 //|
```
Comment (dhalbert) [inline shared-bindings/displayio/Group.c]: Need to document `:param int scale` here.

File: shared-bindings/displayio/Group.c [hunk 2]
```diff
@@ -41,57 +41,236 @@
...
         { MP_QSTR_max_size, MP_ARG_INT | MP_ARG_KW_ONLY, {.u_int = 4} },
+        { MP_QSTR_scale, MP_ARG_INT | MP_ARG_KW_ONLY, {.u_int = 1} },
     };
     mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
     mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
 
     mp_int_t max_size = args[ARG_max_size].u_int;
     if (max_size < 1) {
-        mp_raise_ValueError(translate("Group must have size at least 1"));
+        mp_raise_ValueError_varg(translate("Group must have %q at least 1"), MP_QSTR_max_size);
```
Comment (dhalbert) [inline shared-bindings/displayio/Group.c]: How about `"%q must be >= 1"` ?

File: shared-bindings/displayio/Group.c [hunk 3]
```diff
@@ -41,57 +41,236 @@
...
+    }
+
+    mp_int_t scale = args[ARG_scale].u_int;
+    if (scale < 1) {
+        mp_raise_ValueError_varg(translate("Group must have %q at least 1"), MP_QSTR_scale);
```
Comment (dhalbert) [inline shared-bindings/displayio/Group.c]: How about `"%q must be >= 1"` ?

File: shared-bindings/displayio/Group.c [hunk 4]
```diff
@@ -41,57 +41,236 @@
...
-    common_hal_displayio_group_construct(self, max_size);
+    common_hal_displayio_group_construct(self, max_size, scale);
 
     return MP_OBJ_FROM_PTR(self);
 }
 
+// Helper to ensure we have the native super class instead of a subclass.
+static displayio_group_t* native_group(mp_obj_t group_obj) {
+    mp_obj_t native_group = mp_instance_cast_to_native_base(group_obj, &displayio_group_type);
+    assert(native_group == MP_OBJ_NULL);
```
Comment (dhalbert) [inline shared-bindings/displayio/Group.c]: Does this raise a Python exception if it fails?

File: shared-bindings/displayio/Group.c [hunk 5]
```diff
@@ -41,57 +41,236 @@
...
+    return MP_OBJ_NEW_SMALL_INT(common_hal_displayio_group_get_scale(self));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(displayio_group_get_scale_obj, displayio_group_obj_get_scale);
+
+STATIC mp_obj_t displayio_group_obj_set_scale(mp_obj_t self_in, mp_obj_t scale_obj) {
+    displayio_group_t *self = native_group(self_in);
+
+    mp_int_t scale = mp_obj_get_int(scale_obj);
+    if (scale < 1) {
+        mp_raise_ValueError_varg(translate("Group must have %q at least 1"), MP_QSTR_scale);
```
Comment (dhalbert) [inline shared-bindings/displayio/Group.c]: How about `"%q must be >= 1"` ?

File: shared-module/displayio/Group.c [hunk 6]
```diff
@@ -29,12 +29,39 @@
...
 }
 
-void common_hal_displayio_group_append(displayio_group_t* self, mp_obj_t layer) {
+uint32_t common_hal_displayio_group_get_scale(displayio_group_t* self) {
+    return self->scale;
+}
+
+void common_hal_displayio_group_set_scale(displayio_group_t* self, uint32_t scale) {
+    self->scale = scale;
+    self->needs_refresh = true;
```
Comment (dhalbert) [inline shared-module/displayio/Group.c]: Check if scale is the same and avoid refresh?

File: shared-module/displayio/Group.c [hunk 7]
```diff
@@ -29,12 +29,39 @@
...
+}
+
+mp_int_t common_hal_displayio_group_get_x(displayio_group_t* self) {
+    return self->x;
+}
+
+void common_hal_displayio_group_set_x(displayio_group_t* self, mp_int_t x) {
+    self->x = x;
+    self->needs_refresh = true;
```
Comment (dhalbert) [inline shared-module/displayio/Group.c]: Check if same and avoid refresh?

File: shared-module/displayio/Group.c [hunk 8]
```diff
@@ -29,12 +29,39 @@
...
+}
+
+mp_int_t common_hal_displayio_group_get_y(displayio_group_t* self) {
+    return self->y;
+}
+
+void common_hal_displayio_group_set_y(displayio_group_t* self, mp_int_t y) {
+    self->y = y;
+    self->needs_refresh = true;
```
Comment (dhalbert) [inline shared-module/displayio/Group.c]: Check if same and avoid refresh?

File: shared-bindings/displayio/Group.c (hunk 4 above)
Comment (tannewt) [reply shared-bindings/displayio/Group.c]: I don't think it does anything. Seems like it should fail. Removed.

**Comment by tannewt** [comment]
@dhalbert Please take another look!

**Comment by tannewt** [comment]
Oops, needed to refresh. /me waits again for Travis

**Pull request merged**